src/seq.rs

```diff
@@ -224,6 +225,57 @@ fn sample_indices_cache<R>(
     out
 }
 
+pub(crate) fn shuffle<R, T>(rng: &mut R, values: &mut [T])
+where R: Rng + ?Sized {
+    // In theory this function is nothing more then:
+    // ```
+    // while i > 1 {
+    //     // invariant: elements with index >= i have been locked in place.
+    //     i -= 1;
+    //     // lock element i in place.
+    //     values.swap(i, self.gen_range(0, i + 1));
+    // }
+    // ```
+    //
+    // We optimize for slices of different, because generating ranges is
+    // faster for smaller integers. Less bits from the RNG are necessary,
+    // and multiplies are faster.
+    //
+    // We don't switch exactly at the boundary between integer sizes,
+    // because right below the integer boundary there is a very large zone
+    // of values that have to be rejected to avoid bias, 25~50%.
+    let mut i = values.len() as u64;
+    while i > (1 << 31) {
+        i -= 1;
+        values.swap(i as usize, rng.gen_range(0, i + 1) as usize);
```

Comment:
We allow sampling `i` here (no swap)

Reply:
If `i` is the length of the slice, `i - 1` is the latest index. We should generate the range over `(i - 1) + 1`, because it is exclusive, and to allow a case where `i` does not get swapped. So I believe this code is correct.

Reply:
In *a single line* you have "swap i and gen_range(0, i+1)", i.e. the latter sample can be i, which is I believe correct.

Reply:
Just confirming :smile: